GafferScene : Add compatibility for scripts with old PathMatcherData

danieldresser-ie opened the pull request:
Not sure if this is appropriate, but we've got production scripts with expressions referencing PathMatcherData the old way ( to access the results of FilterResults ), and something like this is necessary to load them.

johnhaddon commented on startup/GafferScene/pathMatcherCompatibility.py:
```diff
@@ -0,0 +1,42 @@
+##########################################################################
+
+import GafferScene
+import Gaffer
```
Unused import

johnhaddon commented on startup/GafferScene/pathMatcherCompatibility.py:
```diff
@@ -0,0 +1,42 @@
+import GafferScene
+import Gaffer
+
+GafferScene.PathMatcher = IECore.PathMatcher
```
We're missing the `import IECore` for this. I think we're probably getting lucky because the `IECore.ConfigLoader` that loads the startup files [reuses the same contextDict](https://github.com/ImageEngine/cortex/blob/master/python/IECore/ConfigLoader.py#L72) for each startup file. I'd like to fix that, but I wouldn't like to bet it wouldn't break something in the startup sequence at IE - what's your feeling @andrewkaufman? If I made a PR would we give it a go?

andrewkaufman replied:
I have a feeling that is considered a feature by at least one essential and highly configured production software, @ivanimanishi to confirm.

johnhaddon replied:
Yeah, I thought that might be the case. It does mean that errors like the one here can go silently unnoticed though, and then a change somewhere else breaks it.

Perhaps I could make a PR allow `contextDict = None` to be passed, signifying that there is no context to share, and a fresh one should be made per file. Then we could use it in Cortex/Gaffer startup without affecting any other projects...

andrewkaufman replied:
That sounds fine

danieldresser-ie commented:
I've added the missing IECore import.

The Gaffer import is needed, because that's where PathMatcherDataPlug lives now.

I've updated the testTypeNamePrefixes - the one really confusing thing is that there was already an "IECore::PathMatcherData" in namesToIgnore - this is confusing for two reasons:
* Things in IECore don't have a module prefix
* What was this code doing back in 2013, since the move of PathMatcherData to IECore happened recently?

Currently, I've just removed "IECore::PathMatcherData", and added "PathMatcherData" and "Gaffer::PathMatcherDataPlug", which is what is currently needed for this to work.  Let me know if there's some reason I should add back "IECore::PathMatcherData".

johnhaddon commented:
> I've added the missing IECore import.

Thanks - merged.

> The Gaffer import is needed, because that's where PathMatcherDataPlug lives now.

Of course - should have spotted that.

> I've updated the testTypeNamePrefixes - the one really confusing thing is that there was already an "IECore::PathMatcherData" in namesToIgnore - this is confusing for two reasons:
>
>   - Things in IECore don't have a module prefix
>   - What was this code doing back in 2013, since the move of PathMatcherData to IECore happened recently?

I don't recall the details, but PathMatcherData specialises IECore::TypedData, so even though it was in GafferScene, the specialisations had to be declared in `namespace IECore`. That either forced me into the "IECore::" prefix, or I made the mistake of adding it.

As an aside, I really think we should prefix the typenames in Cortex too.

> Currently, I've just removed "IECore::PathMatcherData", and added "PathMatcherData" and "Gaffer::PathMatcherDataPlug", which is what is currently needed for this to work. Let me know if there's some reason I should add back "IECore::PathMatcherData".

Nope - that's spot on. Thanks...
